### PR TITLE
[scripts|world rebuild] Always save `config.log` and `CMakeCache.txt`

### LINKED ISSUE #21422: autotools configure config.log not preserved
In case the ./configure step fails, the config.log file containing the details is not preserved. 
This happens because a FATAL_ERROR is issue before the copy of the file is done here.  
https://github.com/microsoft/vcpkg/blob/d369df7ecf194005eaca46f07368779cd486badd/scripts/cmake/vcpkg_execute_required_process.cmake#L127
The move statement lives here:
https://github.com/microsoft/vcpkg/blob/c973b499f2ec648d9e7ae6c784f94d0ebc41f788/scripts/cmake/vcpkg_configure_make.cmake#L811

It works in the good case, where the info is not needed, but it fails in the failure case where this info is interesting the most. 

**Proposed solution**
As a workaround I have made the error non fatal. 

**Describe alternatives you've considered**
Not sure how is the best way around the issue. Maybe we can tell vcpkg_execute_required_process to save the config.log file on failure.

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?
  Reliably saves `config.log` (from `configure`) and `CMakeCache.txt` (from CMake) for build failure analysis.
  Fixes https://github.com/microsoft/vcpkg/issues/21422. CC @daschuer.
  ~This particular implementation tries to avoid forcing a total world rebuild by only touching `vcpkg_configure_make`. For alternatives ideas, see https://github.com/microsoft/vcpkg/pull/25188#issuecomment-1152944087.~

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  all, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  TBD (`vcpkg-cmake`)
  
The simulated error change is a proof of concept and needs to reverted.

### CODE REVIEW

**Comment by dg0yt** [comment]
Not yet properly forwarding arguments.

**Comment by dg0yt** [comment]
Looks like a better start this time. Macros are unable to forward certain arguments, here: the literal `${prefix}`.

**Comment by Neumann-A** [comment]
There is an open PR in vcpkg-tool which tries to find all log files in the buildtree and add them to the CI output. It is not just `config.log` which is the problem but also `CMake(Error|Output).log` or meson logs. Since this is more a hack than a solution I don't think this is mergeable.

**Comment by dg0yt** [comment]
> There is an open PR in vcpkg-tool which tries to find all log files in the buildtree and add them to the CI output. It is not just `config.log` which is the problem but also `CMake(Error|Output).log` or meson logs. Since this is more a hack than a solution I don't think this is mergeable.

 - https://github.com/microsoft/vcpkg-tool/pull/226: I'm aware of that PR. 
   Unfortunately, there is no reasonable estimate for anything being completed.
   That PR is open since Oct 2021.
 -  #21422 is open since Nov 2021.
 - #9802 is open since Oct 2020
 - "CI output": Indeed, https://github.com/microsoft/vcpkg-tool/pull/226 is implemented in the `vcpkg ci` command.
   But for local work, I want consistent behaviour also for `vcpkg install`, for maintainers and for users.
 - Instead of blindly collection all logs for CI, it might make sense to collect named files. ATM there is a questionable pattern of naming things `.log` just to have them collected.
 - I would actually prefer a general hook into tear-down after (failed) `execute_process`. There is not just collecting the logs, but also extracting key messages from the logs and bringing them to the console, or taking other action. Which depends on the actual type of process (download, autotools configure, meson, build/OOM, pkg-config fixup, etc.).
 
So take my PR as an exploration of what is feasible for this particular problem without a total world rebuild or tool changes.
And of course it would be nice if there was a space for such changes which is separate from the ABI-hashed space.

**Comment by daschuer** [comment]
I can confirm that this keeps the config.log as desired: 
https://github.com/daschuer/vcpkg/actions/runs/2481575616

```

-- Configuring x64-osx-dynamic-min10.12-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:128 (message):
    Command failed: /bin/bash -c "V=1 ././configure  \"--disable-nls\" \"--with-openssl\" \"--with-zlib\" \"--without-icu\" \"--with-lz4\" \"--without-readline\" \"--without-libxml\" \"--without-libxslt\" \"--without-python\" \"--disable-silent-rules\" \"--verbose\" \"--enable-shared\" \"--disable-static\" \"--enable-debug\" \"--prefix=/Users/runner/mixxx-vcpkg/installed/x64-osx-dynamic-min10.12/debug\" \"--bindir=\\${prefix}/../tools/libpq/debug/bin\" \"--sbindir=\\${prefix}/../tools/libpq/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libpq\""
    Working Directory: /Users/runner/mixxx-vcpkg/buildtrees/libpq/x64-osx-dynamic-min10.12-dbg
    Error code: 1
    See logs for more information:
      /Users/runner/mixxx-vcpkg/buildtrees/libpq/config-x64-osx-dynamic-min10.12-dbg-out.log
      /Users/runner/mixxx-vcpkg/buildtrees/libpq/config-x64-osx-dynamic-min10.12-dbg-err.log
Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:723 (vcpkg_execute_required_process)
  ports/libpq/portfile.cmake:288 (vcpkg_configure_make)
  scripts/ports.cmake:146 (include)
```

This is a very welcome addition. Thank you.

**Comment by dg0yt** [comment]
> I can confirm that this keeps the config.log as desired:

There is also the simulated `elfutils` error in this PR, making the desired behaviour visible in the failure logs from vcpkg CI.

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 1]
```diff
@@ -641,6 +641,22 @@ function(vcpkg_configure_make)
...
 
+    if(NOT COMMAND _vcpkg_execute_in_download_mode)
+        # Always save config.log to CURRENT_BUILDTREES_DIR, even on failure.
+        function(vcpkg_execute_in_download_mode)
+            _vcpkg_execute_in_download_mode(${ARGV})
+            set(output_variables "RESULT_VARIABLE;RESULTS_VARIABLE;OUTPUT_VARIABLE;ERROR_VARIABLE")
+            cmake_parse_arguments(PARSE_ARGV 0 arg "" "WORKING_DIRECTORY;${output_variables}" "")
+            if(arg_WORKING_DIRECTORY AND EXISTS "${arg_WORKING_DIRECTORY}/config.log")
+                get_filename_component(suffix "${arg_WORKING_DIRECTORY}" NAME)
+                file(RENAME "${arg_WORKING_DIRECTORY}/config.log" "${CURRENT_BUILDTREES_DIR}/config.log-${suffix}.log")
```
Comment (dg0yt) [inline scripts/cmake/vcpkg_configure_make.cmake:652]: I would prefer copying over renaming.

**Comment by BillyONeal** [comment]
> * [Collect failure logs recursively vcpkg-tool#226](https://github.com/microsoft/vcpkg-tool/pull/226): I'm aware of that PR.
>   Unfortunately, there is no reasonable estimate for anything being completed.
>   That PR is open since Oct 2021.

The problem with this PR remains that it ends up globbing entire source trees, so it seems unlikely that we would take it in its current form. We would want some way for ports to tell the tool "I also put log files over here", and have the tool respond by packaging those up.

If we want to do an intermediate step here, we could add a new parameter to `vcpkg_execute_in_download_mode` (rather than overriding it, as is done here), which is a set of logs to copy out even on failure?

/cc @ras0219-msft @ras0219

**Comment by dg0yt** [comment]
> If we want to do an intermediate step here, we could add a new parameter to `vcpkg_execute_in_download_mode` (rather than overriding it, as is done here), which is a set of logs to copy out even on failure?

This is close to my alternative proposal. The extra parameter would need to be added to `vcpkg_execute_required_process`. This is the function which throws the error.
And just copying the files doesn't help with extra processing of these logs. There are contexts where I would like to process the logs before exiting. Couldn't this function take an optional `RESULT_VARIABLE` parameter which would return the error to the caller instead of raising `FATAL_ERROR`?

**Comment by dg0yt** [comment]
One of my alternative proposals is a tear-down or error-handling function hook. In this regard, look at these PRs:
-  #25173: The "compilation environment" code is an example of what is generally implemented `vcpkg_execute required_process` but is relevant only for particular use cases and environments (MSVC).
- #17505: An attempt to deal with the "compilation environment" for MinGW, at the stage of `detect_compiler`.

**Comment by dg0yt** [comment]
Update:
The copying is now implemented directly in `vcpkg_execute_required_process`. It is not strictly necessary to add a parameter: We can simply always look for `config.log`. (But it is rebuilding the world now anyways.)
I added the same behaviour for `CMakeCache.txt`, to offer extra insights for failed cmake ports.
The resulting file names are a concatenation of given log name prefix (carrying step, triplet, build type) and original file name. If missing, `.log` is appended, for CI log collection.

**Comment by dg0yt** [comment]
Output for elfutils:
~~~
    See logs for more information:
      /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/elfutils/config-x64-linux-rel-config.log
      /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/elfutils/config-x64-linux-rel-out.log
      /home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/elfutils/config-x64-linux-rel-err.log
~~~

**Comment by dg0yt** [comment]
> We can simply always look for `config.log`.

Maybe not the best idea for successful build: it will copy log for each step, once created.

**Comment by dg0yt** [comment]
> If we want to do an intermediate step here, we could add a new parameter to ~`vcpkg_execute_in_download_mode`~`vcpkg_execute_required_process` (rather than overriding it, as is done here), which is a set of logs to copy out even on failure?

This is implemented now. Please review.

**Review by JackBoosY — DISMISSED** [review]
LGTM. Please remove the debug flags.

File: ports/elfutils/portfile.cmake [hunk 2]
```diff
@@ -25,6 +25,7 @@ vcpkg_configure_make(
             --enable-maintainer-mode
     OPTIONS_RELEASE
             ac_cv_null_dereference=no # deactivating Werror due to null dereferences since NDEBUG is passed and asserts thus disabled/removed
+            CC=false
```
Comment (daschuer) [inline ports/elfutils/portfile.cmake:28]: Debug leftover

File: ports/cpr/portfile.cmake [hunk 3]
```diff
@@ -24,6 +24,7 @@ vcpkg_cmake_configure(
     OPTIONS_DEBUG
         -DDISABLE_INSTALL_HEADERS=ON
 )
+message(FATAL_ERROR "STOP")
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/cpr/portfile.cmake:27]: No, a proof of concept (CI failure logs.)

**Review by JackBoosY — CHANGES_REQUESTED** [review]
Please bump the port version. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).
Also please run command `vcpkg x-add-version --all` then commit the changes.

**Review by JackBoosY — APPROVED** [review]

**Review by BillyONeal — APPROVED** [review]

**Comment by JackBoosY** [comment]
Finally, we can get all the required logs in our pipeline test results.

**Comment by dg0yt** [comment]
And I just benefit from the saved CMakeCache.txt: #25645.

**Pull request merged**